File: scripts/convert-pig-commercial.py
"""
Conversion Script: Pig Commercial Assessment Data
Converts Pig_commercial_final.json to system-compatible format
"""

import json
import os
from pathlib import Path

# Paths
SCRIPT_DIR = Path(__file__).parent
INPUT_FILE = SCRIPT_DIR.parent.parent / 'Pig_commercial_final.json'
OUTPUT_DIR = SCRIPT_DIR.parent

# Category to Focus Area mapping
CATEGORY_TO_FOCUS_AREA = {
    'farm_information': 1,
    'breeding_animal_supply': 1,
    'feed_water_supply': 2,
    'visitor_management': 2,
    'equipment_vehicles': 2,
    'disease_management': 3,
    'manure_carcass': 3,
    'vermin_control': 3,
    'production_practices': 3,
    'cleaning_disinfection': 4,
    'maintenance': 4
}

# Focus Area metadata
FOCUS_AREA_METADATA = {
    1: {
        'name': {'en': 'Purchase & Transport', 'id': 'Pembelian & Transportasi', 'vt': 'Mua & Vận Chuyển'},
        'description': {
            'en': 'Prevent disease entry through animals and vehicles',
            'id': 'Mencegah masuknya penyakit melalui hewan dan kendaraan',
            'vt': 'Ngăn ngừa bệnh xâm nhập qua động vật và phương tiện'
        },
        'category': 'external_biosecurity',
        'sections': ['A', 'B']
    },
    2: {
        'name': {'en': 'Facilities & People', 'id': 'Fasilitas & Orang', 'vt': 'Cơ Sở & Con Người'},
        'description': {
            'en': 'Control access and maintain biosecure facilities',
            'id': 'Kontrol akses dan pertahankan fasilitas biosekuriti',
            'vt': 'Kiểm soát truy cập và duy trì cơ sở an toàn sinh học'
        },
        'category': 'external_biosecurity',
        'sections': ['C', 'D', 'E']
    },
    3: {
        'name': {'en': 'Production Management', 'id': 'Manajemen Produksi', 'vt': 'Quản Lý Sản Xuất'},
        'description': {
            'en': 'Manage disease risk and production practices',
            'id': 'Kelola risiko penyakit dan praktik produksi',
            'vt': 'Quản lý rủi ro bệnh và thực hành sản xuất'
        },
        'category': 'internal_biosecurity',
        'sections': ['F', 'G', 'H', 'I']
    },
    4: {
        'name': {'en': 'Cleaning & Disinfection', 'id': 'Pembersihan & Disinfeksi', 'vt': 'Vệ Sinh & Khử Trùng'},
        'description': {
            'en': 'Maintain hygiene through proper cleaning and maintenance',
            'id': 'Pertahankan kebersihan melalui pembersihan dan pemeliharaan yang tepat',
            'vt': 'Duy trì vệ sinh thông qua làm sạch và bảo trì đúng cách'
        },
        'category': 'internal_biosecurity',
        'sections': ['J', 'K']
    }
}

def extract_lang_text(obj, lang):
    """Extract language-specific text"""
    if not obj:
        return ''
    if isinstance(obj, str):
        return obj
    
    # Map vt to vi
    lang_key = 'vi' if lang == 'vt' else lang
    return obj.get(lang_key, obj.get('en', ''))

def convert_answer_type(answer_type):
    """Convert answer type"""
    type_map = {
        'single_choice': 'single_choice',
        'multi_choice': 'multiple_choice',
        'number_input': 'numeric',
        'text_input': 'text'
    }
    return type_map.get(answer_type, answer_type)

def convert_conditional_logic(cond_logic):
    """Convert conditional logic format"""
    if not cond_logic:
        return None
    
    for answer, action in cond_logic.items():
        if action.startswith('skip_to_'):
            target_id = action.replace('skip_to_', '').lower()
            return {
                'if_answer': answer,
                'then': 'skip_to',
                'target': target_id
            }
    return None

def infer_farm_type_relevance(question):
    """Infer farm type relevance from question content"""
    question_str = json.dumps(question).lower()
    
    relevance = []
    
    if any(word in question_str for word in ['breeding', 'sow', 'gilt', 'boar', 'farrowing', 'suckling']):
        relevance.extend(['breeding', 'farrow_to_finish'])
    
    if any(word in question_str for word in ['slaughter', 'finishing', 'piglet']):
        relevance.extend(['finishing', 'farrow_to_finish'])
    
    if 'weaned' in question_str or 'nursery' in question_str:
        relevance.extend(['nursery', 'farrow_to_finish'])
    
    if not relevance:
        return ['breeding', 'finishing', 'nursery', 'farrow_to_finish']
    
    return list(set(relevance))

def convert_question(question, lang, question_number, focus_area_number, section_letter):
    """Convert a single question"""
    converted = {
        'number': question_number,
        'id': question['id'].lower(),
        'section': section_letter,
        'focus_area': focus_area_number,
        'text': extract_lang_text(question['question'], lang),
        'type': convert_answer_type(question['answer_type']),
        'required': question.get('required', False)
    }
    
    # Convert options (normalize scoring 0-10 → 0-100)
    if 'options' in question:
        converted['options'] = []
        for opt in question['options']:
            converted['options'].append({
                'value': opt['id'],
                'text': extract_lang_text(opt['label'], lang),
                'score': opt.get('score', 0) * 10
            })
    
    # Add validation
    if 'validation' in question:
        converted['validation'] = question['validation']
    
    # Add conditional logic
    cond_logic = convert_conditional_logic(question.get('conditional_logic'))
    if cond_logic:
        converted['conditional_logic'] = cond_logic
    
    # Farm type relevance
    converted['farm_type_relevance'] = infer_farm_type_relevance(question)
    
    # Enhanced risk assessment
    if 'risk_assessment' in question:
        ra = question['risk_assessment']
        converted['risk_assessment'] = {
            'risk_description': extract_lang_text(ra.get('risk_description'), lang),
            'recommendation': extract_lang_text(ra.get('recommendation'), lang),
            'priority': ra.get('priority', 'medium'),
            'trigger_score_threshold': ra.get('trigger_score', 5) * 10,
            'diseases_affected': ra.get('diseases_affected', [])
        }
    
    return converted

def convert_farm_profile(source_data, lang):
    """Convert farm profile questions"""
    farm_info_cat = next((cat for cat in source_data['categories'] if cat['id'] == 'farm_information'), None)
    
    if not farm_info_cat:
        raise ValueError('farm_information category not found')
    
    profile = {
        'title': extract_lang_text(farm_info_cat['name'], lang),
        'description': extract_lang_text(farm_info_cat['description'], lang),
        'total_questions': len(farm_info_cat['questions']),
        'questions': []
    }
    
    roman_numerals = ['I', 'II', 'III', 'IV', 'V', 'VI', 'VII']
    
    for idx, q in enumerate(farm_info_cat['questions']):
        converted = {
            'number': roman_numerals[idx] if idx < len(roman_numerals) else str(idx + 1),
            'id': f"pre_{q['id'].lower()}",
            'text': extract_lang_text(q['question'], lang),
            'type': convert_answer_type(q['answer_type']),
            'required': q.get('required', False)
        }
        
        if 'options' in q:
            converted['options'] = []
            for opt in q['options']:
                converted['options'].append({
                    'value': opt['id'],
                    'text': extract_lang_text(opt['label'], lang)
                })
        
        if 'validation' in q:
            converted['validation'] = q['validation']
        
        cond_logic = convert_conditional_logic(q.get('conditional_logic'))
        if cond_logic:
            converted['conditional_logic'] = cond_logic
        
        converted['farm_type_relevance'] = ['breeding', 'finishing', 'nursery', 'farrow_to_finish']
        
        if q['id'] == 'Q1':
            converted['farm_type_detection'] = True
        
        profile['questions'].append(converted)
    
    return profile

def convert_assessment(source_data, lang):
    """Convert assessment questions"""
    focus_areas = []
    global_question_number = 1
    
    for fa_num in range(1, 5):
        focus_area = {
            'number': fa_num,
            'name': extract_lang_text(FOCUS_AREA_METADATA[fa_num]['name'], lang),
            'description': extract_lang_text(FOCUS_AREA_METADATA[fa_num]['description'], lang),
            'category': FOCUS_AREA_METADATA[fa_num]['category'],
            'sections': FOCUS_AREA_METADATA[fa_num]['sections'],
            'questions': []
        }
        
        section_index = 0
        
        for category in source_data['categories']:
            if CATEGORY_TO_FOCUS_AREA.get(category['id']) == fa_num:
                if category['id'] == 'farm_information':
                    continue
                
                section_letter = FOCUS_AREA_METADATA[fa_num]['sections'][section_index] if section_index < len(FOCUS_AREA_METADATA[fa_num]['sections']) else 'A'
                
                for question in category['questions']:
                    converted = convert_question(
                        question,
                        lang,
                        global_question_number,
                        fa_num,
                        section_letter
                    )
                    focus_area['questions'].append(converted)
                    global_question_number += 1
                
                section_index += 1
        
        focus_area['total_questions'] = len(focus_area['questions'])
        focus_area['estimated_time_minutes'] = max(1, len(focus_area['questions']) // 4)
        
        focus_areas.append(focus_area)
    
    return {
        'total_questions': global_question_number - 1,
        'focus_areas': focus_areas
    }

def convert_glossary(source_data, lang):
    """Convert glossary"""
    glossary = {}
    for term, definition in source_data['glossary'].items():
        glossary[term] = extract_lang_text(definition, lang)
    return glossary

def generate_language_file(source_data, lang):
    """Generate output file for a specific language"""
    print(f"\n🔄 Converting to {lang.upper()}...")
    
    output = {
        'language': lang,
        'language_name': {'en': 'English', 'id': 'Indonesian', 'vt': 'Vietnamese'}[lang],
        'version': source_data.get('version', '4.0'),
        'source': source_data.get('source', 'Biocheck.Gent BV'),
        'glossary': convert_glossary(source_data, lang),
        'farm_profile': convert_farm_profile(source_data, lang),
        'assessment': convert_assessment(source_data, lang)
    }
    
    print(f"   ✅ Farm profile: {len(output['farm_profile']['questions'])} questions")
    print(f"   ✅ Assessment: {output['assessment']['total_questions']} questions across {len(output['assessment']['focus_areas'])} focus areas")
    
    return output

def main():
    print('🚀 Starting Pig Commercial Assessment Conversion\n')
    print('=' * 60)
    
    try:
        # Load input data
        print(f'📂 Loading input file: {INPUT_FILE}')
        
        if not INPUT_FILE.exists():
            raise FileNotFoundError(f'Input file not found: {INPUT_FILE}')
        
        with open(INPUT_FILE, 'r', encoding='utf-8') as f:
            source_data = json.load(f)
        
        print('✅ Loaded successfully')
        print(f"   - Total categories: {len(source_data['categories'])}")
        print(f"   - Total questions: {source_data['total_questions']}")
        
        # Convert for each language
        for lang in ['en', 'id', 'vt']:
            converted = generate_language_file(source_data, lang)
            
            output_path = OUTPUT_DIR / f'questions_{lang}.json'
            
            print(f"\n💾 Writing {lang.upper()} file...")
            print(f"   Path: {output_path}")
            
            # Create backup if exists
            if output_path.exists():
                backup_path = output_path.with_suffix('.json.backup')
                output_path.rename(backup_path)
                print(f"   📋 Backup created: {backup_path}")
            
            # Write new file
            with open(output_path, 'w', encoding='utf-8') as f:
                json.dump(converted, f, indent=2, ensure_ascii=False)
            
            print(f"   ✅ File written successfully")
        
        print('\n' + '=' * 60)
        print('✅ Conversion completed successfully!')
        print('\nGenerated files:')
        print(f"   - {OUTPUT_DIR / 'questions_en.json'}")
        print(f"   - {OUTPUT_DIR / 'questions_id.json'}")
        print(f"   - {OUTPUT_DIR / 'questions_vt.json'}")
        print('\n💡 Next steps:')
        print('   1. Review generated files for accuracy')
        print('   2. Test with development server: npm run dev')
        print('   3. Complete a test assessment to verify functionality')
        
    except Exception as error:
        print(f'\n❌ Conversion failed: {error}')
        import traceback
        traceback.print_exc()
        exit(1)

if __name__ == '__main__':
    main()

File: scripts/convert-pig-commercial.js
/**
 * Minimal Conversion Script for Pig Commercial Assessment
 * Converts new JSON format to system-compatible format
 */

import { readFileSync, writeFileSync, existsSync, copyFileSync } from 'fs';
import { join, dirname } from 'path';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);

console.log('🚀 Starting conversion...\n');

try {
    // Step 1: Load input file
    const inputPath = join(__dirname, '..', '..', 'Pig_commercial_final.json');
    console.log('📂 Loading:', inputPath);

    if (!existsSync(inputPath)) {
        throw new Error(`File not found: ${inputPath}`);
    }

    const rawData = readFileSync(inputPath, 'utf8');
    const sourceData = JSON.parse(rawData);

    console.log(`✅ Loaded ${sourceData.total_questions} questions from ${sourceData.categories.length} categories\n`);

    // Step 2: Process each language
    const languages = [
        { code: 'en', name: 'English', viKey: 'en' },
        { code: 'id', name: 'Indonesian', viKey: 'id' },
        { code: 'vt', name: 'Vietnamese', viKey: 'vi' }
    ];

    for (const lang of languages) {
        console.log(`🔄 Processing ${lang.name}...`);

        const output = {
            language: lang.code,
            language_name: lang.name,
            version: '4.0-enhanced',
            source: 'Biocheck.Gent BV',
            glossary: {},
            farm_profile: {
                title: '',
                description: '',
                total_questions: 0,
                questions: []
            },
            assessment: {
                total_questions: 0,
                focus_areas: []
            }
        };

        // Convert glossary
        for (const [term, def] of Object.entries(sourceData.glossary)) {
            output.glossary[term] = typeof def === 'string' ? def : (def[lang.viKey] || def.en);
        }

        // Convert farm profile (first category)
        const farmInfoCat = sourceData.categories.find(c => c.id === 'farm_information');
        if (farmInfoCat) {
            output.farm_profile.title = farmInfoCat.name[lang.viKey] || farmInfoCat.name.en;
            output.farm_profile.description = farmInfoCat.description[lang.viKey] || farmInfoCat.description.en;
            output.farm_profile.total_questions = farmInfoCat.questions.length;

            const romanNumerals = ['I', 'II', 'III', 'IV', 'V', 'VI', 'VII'];

            farmInfoCat.questions.forEach((q, idx) => {
                const converted = {
                    number: romanNumerals[idx] || (idx + 1).toString(),
                    id: `pre_${q.id.toLowerCase()}`,
                    text: q.question[lang.viKey] || q.question.en,
                    type: q.answer_type === 'multi_choice' ? 'multiple_choice' :
                        q.answer_type === 'number_input' ? 'numeric' : q.answer_type,
                    required: q.required || false,
                    farm_type_relevance: ['breeding', 'finishing', 'nursery', 'farrow_to_finish']
                };

                if (q.options) {
                    converted.options = q.options.map(opt => ({
                        value: opt.id,
                        text: opt.label[lang.viKey] || opt.label.en
                    }));
                }

                if (q.validation) {
                    converted.validation = q.validation;
                }

                if (q.conditional_logic) {
                    for (const [answer, action] of Object.entries(q.conditional_logic)) {
                        if (action.startsWith('skip_to_')) {
                            converted.conditional_logic = {
                                if_answer: answer,
                                then: 'skip_to',
                                target: action.replace('skip_to_', '').toLowerCase()
                            };
                            break;
                        }
                    }
                }

                if (q.id === 'Q1') {
                    converted.farm_type_detection = true;
                }

                output.farm_profile.questions.push(converted);
            });
        }

        // Convert assessment - simplified version
        // Focus Area 1: Questions 1-30 (Purchase & Transport)
        // Focus Area 2: Questions 31-60 (Facilities & People)
        // Focus Area 3: Questions 61-90 (Production Management)
        // Focus Area 4: Questions 91-116 (Cleaning & Disinfection)

        const focusAreaMeta = [
            {
                num: 1, name: { en: 'Purchase & Transport', id: 'Pembelian & Transportasi', vi: 'Mua & Vận Chuyển' },
                desc: { en: 'Prevent disease entry through animals and vehicles', id: 'Mencegah masuknya penyakit melalui hewan dan kendaraan', vi: 'Ngăn ngừa bệnh xâm nhập qua động vật và phương tiện' },
                category: 'external_biosecurity', sections: ['A', 'B']
            },
            {
                num: 2, name: { en: 'Facilities & People', id: 'Fasilitas & Orang', vi: 'Cơ Sở & Con Người' },
                desc: { en: 'Control access and maintain biosecure facilities', id: 'Kontrol akses dan pertahankan fasilitas biosekuriti', vi: 'Kiểm soát truy cập và duy trì cơ sở an toàn sinh học' },
                category: 'external_biosecurity', sections: ['C', 'D', 'E']
            },
            {
                num: 3, name: { en: 'Production Management', id: 'Manajemen Produksi', vi: 'Quản Lý Sản Xuất' },
                desc: { en: 'Manage disease risk and production practices', id: 'Kelola risiko penyakit dan praktik produksi', vi: 'Quản lý rủi ro bệnh và thực hành sản xuất' },
                category: 'internal_biosecurity', sections: ['F', 'G', 'H', 'I']
            },
            {
                num: 4, name: { en: 'Cleaning & Disinfection', id: 'Pembersihan & Disinfeksi', vi: 'Vệ Sinh & Khử Trùng' },
                desc: { en: 'Maintain hygiene through proper cleaning and maintenance', id: 'Pertahankan kebersihan melalui pembersihan dan pemeliharaan yang tepat', vi: 'Duy trì vệ sinh thông qua làm sạch và bảo trì đúng cách' },
                category: 'internal_biosecurity', sections: ['J', 'K']
            }
        ];

        let globalQNum = 1;

        // Collect all questions from all categories (except farm_information)
        const allQuestions = [];
        for (const cat of sourceData.categories) {
            if (cat.id !== 'farm_information') {
                allQuestions.push(...cat.questions);
            }
        }

        // Distribute questions across 4 focus areas
        const questionsPerFA = Math.ceil(allQuestions.length / 4);

        for (let faIdx = 0; faIdx < 4; faIdx++) {
            const faMeta = focusAreaMeta[faIdx];
            const focusArea = {
                number: faMeta.num,
                name: faMeta.name[lang.viKey] || faMeta.name.en,
                description: faMeta.desc[lang.viKey] || faMeta.desc.en,
                category: faMeta.category,
                sections: faMeta.sections,
                questions: [],
                total_questions: 0,
                estimated_time_minutes: 10
            };

            const startIdx = faIdx * questionsPerFA;
            const endIdx = Math.min((faIdx + 1) * questionsPerFA, allQuestions.length);
            const faQuestions = allQuestions.slice(startIdx, endIdx);

            for (const q of faQuestions) {
                const sectionLetter = faMeta.sections[0] || 'A';

                const converted = {
                    number: globalQNum,
                    id: q.id.toLowerCase(),
                    section: sectionLetter,
                    focus_area: faMeta.num,
                    text: q.question[lang.viKey] || q.question.en,
                    type: q.answer_type === 'multi_choice' ? 'multiple_choice' :
                        q.answer_type === 'number_input' ? 'numeric' : q.answer_type,
                    required: q.required || false,
                    farm_type_relevance: ['breeding', 'finishing', 'nursery', 'farrow_to_finish']
                };

                if (q.options) {
                    converted.options = q.options.map(opt => ({
                        value: opt.id,
                        text: opt.label[lang.viKey] || opt.label.en,
                        score: (opt.score || 0) * 10 // Convert 0-10 to 0-100
                    }));
                }

                if (q.validation) {
                    converted.validation = q.validation;
                }

                if (q.conditional_logic) {
                    for (const [answer, action] of Object.entries(q.conditional_logic)) {
                        if (action.startsWith('skip_to_')) {
                            converted.conditional_logic = {
                                if_answer: answer,
                                then: 'skip_to',
                                target: action.replace('skip_to_', '').toLowerCase()
                            };
                            break;
                        }
                    }
                }

                if (q.risk_assessment) {
                    const ra = q.risk_assessment;
                    converted.risk_assessment = {
                        risk_description: ra.risk_description[lang.viKey] || ra.risk_description.en,
                        recommendation: ra.recommendation[lang.viKey] || ra.recommendation.en,
                        priority: ra.priority || 'medium',
                        trigger_score_threshold: (ra.trigger_score || 5) * 10,
                        diseases_affected: ra.diseases_affected || []
                    };
                }

                focusArea.questions.push(converted);
                globalQNum++;
            }

            focusArea.total_questions = focusArea.questions.length;
            focusArea.estimated_time_minutes = Math.max(5, Math.ceil(focusArea.questions.length / 4));

            output.assessment.focus_areas.push(focusArea);
        }

        output.assessment.total_questions = globalQNum - 1;

        // Write output file
        const outputPath = join(__dirname, '..', `questions_${lang.code}.json`);

        // Backup existing file
        if (existsSync(outputPath)) {
            const backupPath = outputPath + '.backup';
            copyFileSync(outputPath, backupPath);
            console.log(`   📋 Backup: ${backupPath}`);
        }

        writeFileSync(outputPath, JSON.stringify(output, null, 2), 'utf8');
        console.log(`   ✅ Written: ${outputPath}`);
        console.log(`   📊 ${output.farm_profile.questions.length} farm profile + ${output.assessment.total_questions} assessment questions\n`);
    }

    console.log('✅ Conversion completed successfully!\n');
    console.log('Next steps:');
    console.log('  1. Review generated files');
    console.log('  2. Run: npm run dev');
    console.log('  3. Test assessment flow');

} catch (error) {
    console.error('\n❌ Error:', error.message);
    console.error(error.stack);
    process.exit(1);
}
